### PR TITLE
feat(plugins): add datadog tracing plugin

### PR DESCRIPTION
**Pull request by mayocream**
### Description

Add docs for the new Datadog Tracing plugin.

KAG-454

### Checklist 

- [x] Review label added 
- [x] PR pointed to correct branch (`main` for immediate publishing, or a release branch: e.g. `release/gateway-3.2`, `release/deck-1.17`)

[DOCU-2918]: https://konghq.atlassian.net/browse/DOCU-2918?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by colinh-kong** [comment]
Needs a header image
image

**Comment by Guaris** [comment]
@mayocream 

Hey mayo, after the parameters this doc is a copy of the open telemetry doc. Could you write a usage section specific to the datadog tracing plugin?

**Comment by mayocream** [comment]
> @mayocream
> 
> Hey mayo, after the parameters this doc is a copy of the open telemetry doc. Could you write a usage section specific to the datadog tracing plugin?

It was a draft, I updated the PR and rebased it, it's ready for review.

File: app/_hub/kong-inc/datadog-tracing/_index.md [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+For more information, see the [Datadog documentation](https://docs.datadoghq.com/agent/docker/apm/?tab=linux).
+
+### Configure the Datadog Tracing plugin
+
+Enable the plugin:
+
+```bash
+curl -X POST http://<admin-hostname>:8001/plugins \
+    --data "name=Datadog-tracing"  \
```
Comment (fffonion) [inline app/_hub/kong-inc/datadog-tracing/_index.md:119]:
```suggestion
    --data "name=datadog-tracing"  \
```

File: app/_hub/kong-inc/datadog-tracing/_index.md [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+{:.note}
+> **Note**: The Datadog Tracing plugin only works with {{site.base_gateway}}'s `tracing_instrumentations` configuration enabled. It is recommended that you set the `tracing_instrumentations` configuration to `request`, to avoid impacting performance, and only trace requests.
+
+The Datadog Agent is required to run the Datadog Tracing plugin. The Datadog Agent can be installed on the same machine as {{site.base_gateway}} or on a different machine. The Datadog Tracing plugin will send spans to the Datadog Agent over HTTP.
+
+### Set up {{site.base_gateway}}
+
+Enable the OpenTelemetry tracing capability in {{site.base_gateway}}'s configuration:
+
+- `tracing_instrumentations = request`: valid values can be found in the [tracing instrumentations](/gateway/latest/reference/configuration/#tracing_instrumentations) section of the configuration documentation.
```
Comment (lena-larionova) [inline app/_hub/kong-inc/datadog-tracing/_index.md:91]:
```suggestion
- `tracing_instrumentations = request`: Valid values can be found in the [tracing instrumentations](/gateway/latest/reference/configuration/#tracing_instrumentations) section of the configuration documentation.
```

File: app/_hub/kong-inc/datadog-tracing/_index.md [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+- `tracing_sampling_rate = 1.0`: tracing instrumentation sampling rate.
```
Comment (lena-larionova) [inline app/_hub/kong-inc/datadog-tracing/_index.md:92]:
```suggestion
- `tracing_sampling_rate = 1.0`: Tracing instrumentation sampling rate.
```

File: app/_hub/kong-inc/datadog-tracing/_index.md [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+docker run -d -p 8126:8126 \
+              --cgroupns host \
+              --pid host \
+              -v /var/run/docker.sock:/var/run/docker.sock:ro \
+              -v /proc/:/host/proc/:ro \
+              -v /sys/fs/cgroup/:/host/sys/fs/cgroup:ro \
+              -e DD_API_KEY=<api-key> \
+              -e DD_APM_NON_LOCAL_TRAFFIC=true \
+              -e DD_APM_ENABLED=true \
+              gcr.io/datadoghq/agent:latest
```
Comment (lena-larionova) [inline app/_hub/kong-inc/datadog-tracing/_index.md:108]:
```suggestion
docker run -d -p 8126:8126 \
  --cgroupns host \
  --pid host \
  -v /var/run/docker.sock:/var/run/docker.sock:ro \
  -v /proc/:/host/proc/:ro \
  -v /sys/fs/cgroup/:/host/sys/fs/cgroup:ro \
  -e DD_API_KEY=<example-api-key> \
  -e DD_APM_NON_LOCAL_TRAFFIC=true \
   -e DD_APM_ENABLED=true \
   gcr.io/datadoghq/agent:latest
```
No need for such deep indents.

File: app/_hub/kong-inc/datadog-tracing/_index.md [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+Enable the OpenTelemetry tracing capability in {{site.base_gateway}}'s configuration:
+
+- `tracing_instrumentations = request`: valid values can be found in the [tracing instrumentations](/gateway/latest/reference/configuration/#tracing_instrumentations) section of the configuration documentation.
+- `tracing_sampling_rate = 1.0`: tracing instrumentation sampling rate.
+  The tracer samples a fixed percentage of all spans following the sampling rate.
+  This allows you to set the sampling rate to a lower value to reduce the impact of the instrumentation on {{site.base_gateway}}'s proxy performance in production.
+
+### Set up the Datadog Agent
+
```
Comment (lena-larionova) [inline app/_hub/kong-inc/datadog-tracing/_index.md:97]: Needs an intro to the section. Something like:

```suggestion

Set up the Datadog Agent. You can use the following example Docker configuration, substituting the `example-api-key` with your own key:

```

File: app/_hub/kong-inc/datadog-tracing/_index.md [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+              -v /var/run/docker.sock:/var/run/docker.sock:ro \
+              -v /proc/:/host/proc/:ro \
+              -v /sys/fs/cgroup/:/host/sys/fs/cgroup:ro \
+              -e DD_API_KEY=<api-key> \
+              -e DD_APM_NON_LOCAL_TRAFFIC=true \
+              -e DD_APM_ENABLED=true \
+              gcr.io/datadoghq/agent:latest
+```
+
+For more information about installing the Datadog Agent on Docker, see [Datadog's documentation](https://docs.datadoghq.com/agent/docker/apm/?tab=linux). If you want to install another container option you can find the instructions [on this list](https://docs.datadoghq.com/containers/).
```
Comment (lena-larionova) [inline app/_hub/kong-inc/datadog-tracing/_index.md:111]:
```suggestion
For more information about installing the Datadog Agent on Docker, see [Datadog's documentation](https://docs.datadoghq.com/agent/docker/apm/?tab=linux). If you want to install another container option, you can find the instructions in [Datadog's container list](https://docs.datadoghq.com/containers/).
```
We should use less generic link titles for better SEO and accessibility.

File: app/_hub/kong-inc/datadog-tracing/_index.md [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+
+```bash
+curl -X POST http://<admin-hostname>:8001/plugins \
+    --data "name=datadog-tracing"  \
+    --data "config.endpoint=http://localhost:8126/v1/traces" \
+    --data "config.service_name=kong" \
+    --data "config.environment=prod"
+```
+
+## OpenTelemetry Functionality
```
Comment (lena-larionova) [inline app/_hub/kong-inc/datadog-tracing/_index.md:125]:
```suggestion
## OpenTelemetry functionality
```

File: app/_hub/kong-inc/datadog-tracing/_index.md [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+- `http.method`: HTTP method
+- `http.url`: HTTP URL
+- `http.host`: HTTP host
+- `http.scheme`: HTTP scheme (http or https)
+- `http.flavor`: HTTP version
+- `net.peer.ip`: Client IP address
+
+### Propagation
+
+The OpenTelemetry plugin propagates the following headers:
```
Comment (lena-larionova) [inline app/_hub/kong-inc/datadog-tracing/_index.md:144]:
```suggestion
The Datadog Tracing plugin propagates the following headers:
```

File: app/_hub/kong-inc/datadog-tracing/_index.md [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+- `Datadog`: [Datadog headers](https://docs.datadoghq.com/tracing/agent/propagation/)
+
+The plugin detects the propagation format from the headers and will use the appropriate format to propagate the span context.
+If no appropriate format is found, the plugin will fallback to the default format, which is `Datadog`.
+
+## Troubleshooting
+
+The Datadog Tracing spans are printed to the console when the log level is set to `debug` in the Kong configuration file.
+
+An example of debug logs output:
```
Comment (lena-larionova) [inline app/_hub/kong-inc/datadog-tracing/_index.md:158]:
```suggestion
Here's an example of the debug log output:
```

File: app/_hub/kong-inc/datadog-tracing/_index.md [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+Span #2 name=rewrite phase: datadog-tracing duration=0.391936ms
+Span #3 name=router duration=0.013824ms
+Span #4 name=access phase: cors duration=1500.824576ms
+Span #5 name=cors: heavy works duration=1500.709632ms attributes={"username":"kongers"}
+Span #6 name=balancer try #1 duration=0.99328ms attributes={"net.peer.ip":"104.21.11.162","net.peer.port":80}
+```
+
+### Known issues
+
+- Datadog tracing Only supports the HTTP protocols (http/https) of {{site.base_gateway}}.
```
Comment (lena-larionova) [inline app/_hub/kong-inc/datadog-tracing/_index.md:172]:
```suggestion
- Datadog Tracing only supports the HTTP protocols (`http`/`https`) of {{site.base_gateway}}.
```

File: app/_hub/kong-inc/datadog-tracing/_index.md [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+- May impact the performance of {{site.base_gateway}}.
```
Comment (lena-larionova) [inline app/_hub/kong-inc/datadog-tracing/_index.md:173]:
```suggestion
- This plugin may impact the performance of {{site.base_gateway}}.
```

File: app/_hub/kong-inc/datadog-tracing/_index.md [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+  It's recommended to set the sampling rate (`tracing_sampling_rate`)
```
Comment (lena-larionova) [inline app/_hub/kong-inc/datadog-tracing/_index.md:174]:
```suggestion
  We recommend setting the sampling rate (`tracing_sampling_rate`)
```

File: app/_hub/kong-inc/datadog-tracing/_index.md [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+  via Kong configuration file when using the Datadog Tracing plugin.
```
Comment (lena-larionova) [inline app/_hub/kong-inc/datadog-tracing/_index.md:175]:
```suggestion
  in the {{site.base_gateway}} configuration file when using the Datadog Tracing plugin.
```

File: app/_hub/kong-inc/datadog-tracing/versions.yml [hunk 14]
```diff
@@ -0,0 +1,7 @@
+strategy: gateway
+
+releases:
+  minimum_version: '3.2.x'
+
+overrides:
+  3.2.x: 0.1.0
```
Comment (lena-larionova) [inline app/_hub/kong-inc/datadog-tracing/versions.yml:7]: Why do we need an override for this plugin version? As far as I know, the plugin version _should_ match the Kong version starting in 3.0. If it doesn't, then something's wrong.

**Review by lena-larionova — CHANGES_REQUESTED** [review]

**Review by lena-larionova — APPROVED** [review]

**Review by lena-larionova — APPROVED** [review]

File: app/_data/tables/plugin_index.yml [hunk 15]
```diff
@@ -439,6 +439,14 @@
       network_config_opts: All
       notes: --
       url: /hub/kong-inc/datadog
+    
+    - name: Datadog Tracing
+      free: No
+      plus: Yes
```
Comment (Guaris) [inline app/_data/tables/plugin_index.yml:445]:
```suggestion
      plus: No
```

**Review by Guaris — APPROVED** [review]

**Pull request merged**